crates/transport-ipc/src/mock.rs

```diff
@@ -43,47 +147,347 @@ impl Default for MockIpcServer {
+    /// Create a new handle to control this server.
+    /// The handle can be used to add responses and trigger shutdown.
+    ///
+    /// # Returns
+    /// * A new `MockIpcHandle` instance
+    pub fn handle(&self) -> MockIpcHandle {
+        MockIpcHandle { inner: self.inner.clone() }
     }
 
-    /// Get the path to the socket.
-    pub fn path(&self) -> PathBuf {
-        self.path.path().to_owned()
+    /// Handle a single client connection.
+    /// This function runs in a separate task for each connected client.
+    ///
+    /// # Arguments
+    /// * `stream` - The Unix domain socket stream for the client
+    /// * `inner` - Reference to shared server state
+    async fn handle_connection(
+        stream: tokio::net::UnixStream,
+        inner: Arc<Inner>,
+    ) -> std::io::Result<()> {
+        let (read, mut writer) = stream.into_split();
+        let mut reader = ReadJsonStream::new(read);
+
+        // Channel for sending responses back to the writer
+        let (tx, mut rx) = mpsc::channel::<Vec<u8>>(32);
+
+        // Collection of in-flight request handlers
+        let mut request_handlers = FuturesUnordered::new();
+
+        debug!("Starting connection handler loop");
+        loop {
+            tokio::select! {
+                // Handle incoming requests
+                maybe_request = reader.next() => {
+                    match maybe_request {
+                        Some(request) => {
+                            if let Ok(request) = serde_json::from_value::<Value>(request) {
+                                debug!(
+                                    id = ?request.get("id"),
+                                    method = ?request.get("method"),
+                                    "Received JSON-RPC request"
+                                );
+
+                                // Clone what we need for the task
+                                let inner = inner.clone();
+                                let tx = tx.clone();
+
+                                // Spawn a new task to handle this request
+                                request_handlers.push(tokio::spawn(async move {
+                                    // Get the next queued response or generate an error
+                                    let response = inner.replies.lock().pop_front().map_or_else(|| {
+                                        warn!("No queued response available for request");
+                                        serde_json::to_vec(&json!({
+                                            "jsonrpc": "2.0",
+                                            "id": request.get("id"),
+                                            "error": {
+                                                "code": -32603,
+                                                "message": "No response queued"
+                                            }
+                                        })).expect("JSON serialization cannot fail")
```

Comment:
imo we can get rid of this additional roundtrip entirely.
we don't need this and can write the response right away.